### DISCUSSION TITLE
[Feature] Support MPO files

### DISCUSSION THREAD

**Discussion by Piipperi** [immich-app/immich#9886] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I don't know if anything uses this format anymore, but the Nintendo 3DS-line appears to save 3D images in MPO format, and I have a couple of those in my iCloud Photos that fail to upload to Immich.

Unsure if this is viable to add as a supported file format, but posting this here as a request either way.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by mjsir911** [comment]
Would appreciate this!

**Comment by genebean** [comment]
I have some of these in iCloud that were taken with a digital camera.

**Comment by WoOS2** [comment]
I support this. I have lots of "3D" (i.e stereoscopic) photos from my old Fuji FinePix Real 3D W3 camera, all in .mpo. 
I think showing them stereoscopically would be complicated (depends on the monitor after all), showing them in 2D should be rather easy given that **.mpo is basically 2 JPEGs**.
So just show (and analyze) the left (or the right) jpeg and only provide the .mpo file on download.
Docs on .mpo:
https://de.wikipedia.org/wiki/Multi_Picture_Object (if you speak German, otherwise: https://de-m-wikipedia-org.translate.goog/wiki/Multi_Picture_Object?_x_tr_sl=de&_x_tr_tl=en&_x_tr_hl=en&_x_tr_pto=wapp).

**Comment by T0biasCZe** [comment]
> I think showing them stereoscopically would be complicated

it could be shown as animation, alternating between the left and right eye, like this:         
![HNI_0063 aligned rotated](https://github.com/user-attachments/assets/687c1d58-b5cf-4207-bf13-dfefcd469146)

**Comment by WoOS2** [comment]
Showing both pictures within the MPO file would be advanced functionality as that will probably play havoc with the thumbnail generation. 
IMHO taking one of the jpegs within the MPO file and creating all the thumbnails from it would be enough. People who really want to see the 3D (or the animation) can download the original.

BTW, greetings to Czechia or Slovakia. I tried that drink and I have to admit to me it taste like medicine.

**Reply by genebean** [reply]
I’d be fine with a single thumbnail as at least it wouldn’t error out when backing things up.

**Comment by T0biasCZe** [comment]
> functionality as that will probably play havoc with the thumbnail generation.

well, gifs currently also show just 1 frame in the thumbnail, and it plays animated only when you open it in fullscreen, so thumbnail gen shouldnt be problem...      
the animated mpo could work the same way. On thumbnail it would show only 1 eye, and when you open it, it would animate it

**Comment by DokiDerg** [comment]
Sucks this still hasnt been added yet, i have 3ds photos as well, and recently got the stuff for a NAS

I actually found how to convert them into 2 jpg so i can make the steroscopic image on my pc, heres some instructions i made to myself if anyone wants to do it manually or make a program (i probably will for this reason)

1. Find 'Exif  MM' (45 78 69 66 00 00 4D 4D)
2. each Exif is one jpg, with 6 bytes before it being the start
3. Take one image as from the 0 byte, to 6 bytes before the second Exif  MM
4. then a second image, 6 bytes before the second Exif  MM
5. This will create 2 JPG with both perspectives

**Comment by T0biasCZe** [comment]
@DokiDerg A program already exists for taking .mpo images, and exporting them to separate jpegs, or even animated gifs
https://stereo.jpn.org/eng/stphmkr/

**Comment by T0biasCZe** [comment]
But yeah it sucks, I have lot of photos from 3ds'es and I too cant back them up.      
It would be enough if Immich even just allow the upload of the .mpo files for archival, without any processing and transcode, since they are just fancy jpegs and browsers can load .mpo as is. (albeit in 2d)

**Comment by DokiDerg** [comment]
I would like an option to upload any file type, even if it cant be processed, maybe in a sepcific tab or filter to show them

**Reply by hurenkam** [reply]
Indeed, this would allow for immich to be used as cloud storage for any media file, regardless if the type is supported. In the viewer you can provide an option to download unsupported media types. If some smartphone or camera manufacturer invents a new media type, it will still get saved to the cloud storage, and you don't have to think of another way to get those media off your device into the cloud. If someone integrates a new viewer for these media in the future, then old media also become automatically viewable, and don't need to be uploaded again.
I do have about 2500+ mpo files in my collection, mostly taken with an HTC EVO 3D back when stereo camera's were not uncommon. Same format is used by Nintendo 3DS, which I still use sometimes.

**Comment by djbiti1** [comment]
That would be actually very useful, especially since immich preserves the original file without altering it. Very useful for archiving. And maybe they will implement a viewer for these MPOs in the future to be able to display them.

**Comment by lethyro85** [comment]
I use mpo in older 3D pictures. I need also this MPO feature :)

**Comment by theS1LV3R** [comment]
If anyone could guide me on where in the codebase I'd have to go to implement this I'd be happy trying to create a PR. Got some MPO files I'd like to back up myself.

**Comment by ejegg** [comment]
I'd be interested in this as well - I've got a ton of pics from the same Fuji FinePix camera mentioned by WoOS2. I created a plugin for piwigo to display them as wiggle gifs or side-by side and would love to see it work in immich

https://ejegg.com/stereodemo/

The code = GPL 2.0 but I'm happy to dual license as AGPL 3 if needed to incorporate anything here : https://github.com/ejegg/piwigo-stereo

**Comment by lethyro85** [comment]
Hi,
i have a demo at https://londor.eu/mpo/mpo_demo.php?file=https://londor.eu/mpo/DSCF6222.MPO
and the code in PHP is here: https://londor.eu/mpo/mpo_demo.txt

it make mpo  to 2 images and to an anaglyph. 
Sorry my comments there is in German :) 

use it :)

**Comment by royord** [comment]
Also very interested in this. When the world lost Aperture by Apple, years ago now, we also lost support for archiving MPO files. Been looking ever since.

**Comment by bo0tzz** [comment]
#27963
